crates/provider-sdk/src/provider.rs

```diff
@@ -196,109 +212,134 @@ impl ProviderConnection {
-                                    current.record("lattice_id", &tracing::field::display(&lattice));
-                                    current.record("actor_id", &tracing::field::display(&inv.origin.public_key));
-                                    current.record("inv_id", &tracing::field::display(&inv.id));
-                                    current.record("host_id", &tracing::field::display(&inv.host_id));
-                                    current.record("provider_id", &tracing::field::display(&inv.target.public_key));
-                                    current.record("contract_id", &tracing::field::display(&inv.target.contract_id));
-                                    current.record("link_name", &tracing::field::display(&inv.target.link_name));
-                                    current.record("payload_size", &tracing::field::display(&inv.content_length));
-                                    let inv_id = inv.id.clone();
-                                    let inv_operation = inv.operation.clone();
-                                    let resp = match this.handle_rpc(provider.clone(), inv).in_current_span().await {
-                                        Err(err) => {
-                                            error!(%err, operation = %inv_operation, "Invocation failed");
-                                            InvocationResponse{
-                                                invocation_id: inv_id,
-                                                error: Some(format!("Error when handling invocation: {err}")),
-                                                ..Default::default()
+        let mut subs = vec![
+            self.rpc_client
+                .client()
+                .queue_subscribe(
+                    self.provider_rpc_topic(),
+                    RPC_SUBSCRIPTION_QUEUE_GROUP.to_string(),
+                )
+                .await?,
+        ];
+
+        // If we have a WIT resolve present, then this is a WIT-driven provider, and we can generate wprc
+        // topics on which we should listen to for RPCs
+        if let Some(resolve) = self.wit_resolve.as_ref() {
+            for topic in generate_incoming_rpcs_topics(&lattice, WRPC_VERSION, resolve)? {
+                subs.push(
+                    self.rpc_client
+                        .client()
+                        .queue_subscribe(topic, RPC_SUBSCRIPTION_QUEUE_GROUP.to_string())
+                        .await?,
+                )
+            }
+        }
+
+        // For every subscription, spawn a task to listen to incoming RPCs received,
+        // keeping track of the handles generated
+        let mut handles = Vec::new();
+        for mut sub in subs {
+            let this = self.clone();
+            let mut quit = quit.resubscribe();
+            let lattice = lattice.clone();
+            let provider = provider.clone();
+            let handle = tokio::spawn(async move {
```

Comment:
Nit: It might be high time to refactor this out into a function we can spawn instead

Reply:
Yeah let's do this in another ticket -- I'd like to focus on just the necessary changes without refactoring for now